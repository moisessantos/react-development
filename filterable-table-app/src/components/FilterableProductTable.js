import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ProductTable, SearchBar}  from './';

class FilterableProductTable extends Component {
  constructor(props){
    super(props);
    this.state = {
      filteredProducts: props.products,
      includeOutOfStock: true
    };

    this.filterProduct = this.filterProduct.bind(this);
    this.filterStock = this.filterStock.bind(this);
  }
  render() {
    let products = this.state.filteredProducts;

    if(!this.state.includeOutOfStock) {
      products = products.filter(product => product.stocked);
    }

    return (
      <div className='wrapper'>
        <SearchBar filter={ this.filterProduct } onClick={ this.filterStock } />
        <ProductTable products={ products } />
      </div>
    );
  }

  filterProduct(name) {
    this.setState({
      filteredProducts: this.state.filteredProducts.filter(product => product.name.indexOf(name) !== -1)
    });
  }

  filterStock(){
    this.setState({
      includeOutOfStock: !this.state.includeOutOfStock
    });
  }
}

FilterableProductTable.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape({
    category: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    stocked: PropTypes.bool,
    name: PropTypes.string.isRequired,
  })).isRequired,
};
export default FilterableProductTable;
