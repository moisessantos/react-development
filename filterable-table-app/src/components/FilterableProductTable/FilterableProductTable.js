import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { ProductTable, SearchBar }  from '..';

class FilterableProductTable extends PureComponent {
  constructor(props){
    super(props);
    this.state = {
      filteredProducts: props.products,
      includeOutOfStock: true
    };
  }
  render() {
    let { filteredProducts, includeOutOfStock } = this.state;

    if(!includeOutOfStock) {
      filteredProducts = filteredProducts.filter(product => product.stocked);
    }

    return (
      <div className='wrapper'>
        <SearchBar handleFilter={ this.filterProduct.bind(this) } handleClick={ this.filterStock.bind(this) } />
        <ProductTable products={ filteredProducts } />
      </div>
    );
  }

  filterProduct(name) {
    this.setState({
      filteredProducts: this.props.products.filter(product => product.name.indexOf(name) !== -1)
    });
  }

  filterStock(){
    this.setState({
      includeOutOfStock: !this.state.includeOutOfStock
    });
  }
}

FilterableProductTable.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape({
    category: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    stocked: PropTypes.bool,
    name: PropTypes.string.isRequired,
  })).isRequired,
};
export default FilterableProductTable;
