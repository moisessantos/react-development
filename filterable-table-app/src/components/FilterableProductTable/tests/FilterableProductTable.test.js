import React from 'react';
import { shallow } from 'enzyme';
import FilterableProductTable from '..';

describe('FilterableProductTable component', () => {
  const mockProducts = [
    {category: 'Sporting Goods', price: '$49.99', stocked: true, name: 'Football'},
    {category: 'Sporting Goods', price: '$29.99', stocked: false, name: 'Basketball'},
    {category: 'Electronics', price: '$99.99', stocked: true, name: 'iPod Touch'}
  ];
  it('should render snapshot', () => {
    const wrapper = shallow(<FilterableProductTable products={ mockProducts } />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should filter by name', () => {
    const wrapper = shallow(<FilterableProductTable products={ mockProducts } />);
    wrapper.instance().filterProduct('iPod');

    expect(wrapper.state('filteredProducts')).toEqual([{category: 'Electronics', price: '$99.99', stocked: true, name: 'iPod Touch'}])
    expect(wrapper).toMatchSnapshot();
  });
  it('should filter by stock', () => {
    const wrapper = shallow(<FilterableProductTable products={ mockProducts } />);
    wrapper.instance().filterStock();
    
    expect(wrapper.state('includeOutOfStock')).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });
});
