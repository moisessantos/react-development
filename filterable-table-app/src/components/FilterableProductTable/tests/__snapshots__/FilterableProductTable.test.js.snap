// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableProductTable component should filter by name 1`] = `
<div
  className="wrapper"
>
  <SearchBar
    handleClick={[Function]}
    handleFilter={[Function]}
  />
  <ProductTable
    products={
      Array [
        Object {
          "category": "Electronics",
          "name": "iPod Touch",
          "price": "$99.99",
          "stocked": true,
        },
      ]
    }
  />
</div>
`;

exports[`FilterableProductTable component should filter by stock 1`] = `
<div
  className="wrapper"
>
  <SearchBar
    handleClick={[Function]}
    handleFilter={[Function]}
  />
  <ProductTable
    products={
      Array [
        Object {
          "category": "Sporting Goods",
          "name": "Football",
          "price": "$49.99",
          "stocked": true,
        },
        Object {
          "category": "Electronics",
          "name": "iPod Touch",
          "price": "$99.99",
          "stocked": true,
        },
      ]
    }
  />
</div>
`;

exports[`FilterableProductTable component should render snapshot 1`] = `
<div
  className="wrapper"
>
  <SearchBar
    handleClick={[Function]}
    handleFilter={[Function]}
  />
  <ProductTable
    products={
      Array [
        Object {
          "category": "Sporting Goods",
          "name": "Football",
          "price": "$49.99",
          "stocked": true,
        },
        Object {
          "category": "Sporting Goods",
          "name": "Basketball",
          "price": "$29.99",
          "stocked": false,
        },
        Object {
          "category": "Electronics",
          "name": "iPod Touch",
          "price": "$99.99",
          "stocked": true,
        },
      ]
    }
  />
</div>
`;
